File: onyx-database-tests/test/memory/queries/SelectQueryTest.java
package memory.queries;

import category.InMemoryDatabaseTests;
import com.onyx.exception.EntityException;
import com.onyx.persistence.query.Query;
import com.onyx.persistence.query.QueryCriteria;
import com.onyx.persistence.query.QueryCriteriaOperator;
import com.onyx.persistence.query.QueryOrder;
import entities.AllAttributeForFetch;
import org.junit.Assert;
import org.junit.Test;
import org.junit.experimental.categories.Category;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

/**
 * Created by timothy.osborn on 1/10/15.
 */
@Category({ InMemoryDatabaseTests.class })
public class SelectQueryTest extends memory.base.PrePopulatedDatabaseTest
{

    @Test
    public void testExecuteSelectFields() throws EntityException, InstantiationException, IllegalAccessException
    {
        QueryCriteria criteria = new QueryCriteria("stringValue", QueryCriteriaOperator.STARTS_WITH, "Some");
        Query query = new Query(AllAttributeForFetch.class, Arrays.asList("longValue", "intPrimitive"), criteria);
        List results = manager.executeQuery(query);
        Assert.assertNotNull(results);
        Assert.assertTrue(results.size() == 4);
        Assert.assertTrue(results.get(0) instanceof Map);
        Assert.assertTrue((Long)((Map) results.get(0)).get("longValue") > 0);
    }

    @Test
    public void testSelectOnlyOne() throws EntityException, InstantiationException, IllegalAccessException
    {
        QueryCriteria criteria = new QueryCriteria("stringValue", QueryCriteriaOperator.STARTS_WITH, "Some");
        Query query = new Query(AllAttributeForFetch.class, Arrays.asList("longValue", "intPrimitive"), criteria);
        query.setFirstRow(0);
        query.setMaxResults(1);
        List results = manager.executeQuery(query);

        Assert.assertNotNull(results);
        Assert.assertTrue(results.size() == 1);
        Assert.assertTrue(results.get(0) instanceof Map);

        if(results.get(0) instanceof Long)
        {
            Assert.assertTrue((Long)((Map) results.get(0)).get("longValue") > 0);
        }
        else if(results.get(0) instanceof Integer)
        {
            Assert.assertTrue((Integer)((Map) results.get(0)).get("longValue") > 0);
        }
    }

    @Test
    public void testSelectTwoOrderBy() throws EntityException, InstantiationException, IllegalAccessException
    {
        QueryCriteria criteria = new QueryCriteria("stringValue", QueryCriteriaOperator.STARTS_WITH, "Some");
        Query query = new Query(AllAttributeForFetch.class, Arrays.asList("id","stringValue","longValue", "intPrimitive"), criteria);
        query.setFirstRow(2);
        query.setMaxResults(2);
        query.setQueryOrders(Arrays.asList(new QueryOrder("stringValue"), new QueryOrder("intPrimitive")));

        List results = manager.executeQuery(query);

        Assert.assertNotNull(results);
//        Assert.assertTrue(results.size() == 2);
        Assert.assertTrue(results.get(0) instanceof Map);
        Assert.assertTrue((int)((Map) results.get(0)).get("intPrimitive") == 4);
        Assert.assertTrue((int)((Map) results.get(1)).get("intPrimitive") == 3);
    }

    @Test
    public void testNoSelect() throws EntityException, InstantiationException, IllegalAccessException
    {
        QueryCriteria criteria = new QueryCriteria("stringValue", QueryCriteriaOperator.STARTS_WITH, "Some");
        Query query = new Query(AllAttributeForFetch.class, criteria);
        query.setFirstRow(0);
        query.setMaxResults(2);
        query.setQueryOrders(Arrays.asList(new QueryOrder("stringValue"), new QueryOrder("intPrimitive")));

        List results = manager.executeQuery(query);

        Assert.assertNotNull(results);
        Assert.assertTrue(results.size() == 2);
        Assert.assertTrue(results.get(0) instanceof AllAttributeForFetch);
        Assert.assertTrue(((AllAttributeForFetch)results.get(0)).id.equals("FIRST ONE"));
        Assert.assertTrue(((AllAttributeForFetch)results.get(1)).id.equals("FIRST ONE2") || ((AllAttributeForFetch)results.get(1)).id.equals("FIRST ONE1"));
    }

    @Test
    public void testSelectRelationship() throws EntityException, InstantiationException, IllegalAccessException
    {
        QueryCriteria criteria = new QueryCriteria("stringValue", QueryCriteriaOperator.STARTS_WITH, "Some")
                                              .and("child.someOtherField", QueryCriteriaOperator.STARTS_WITH, "HIYA");
        Query query = new Query(AllAttributeForFetch.class, criteria);
        query.setFirstRow(0);
        query.setMaxResults(2);
        query.setQueryOrders(Arrays.asList(new QueryOrder("child.someOtherField"), new QueryOrder("intPrimitive")));
        query.setSelections(Arrays.asList("longValue", "intPrimitive", "child.someOtherField"));

        List results = manager.executeQuery(query);

        Assert.assertNotNull(results);
        Assert.assertTrue(results.size() == 1);
        Assert.assertTrue(results.get(0) instanceof Map);
    }

    @Test
    public void testSelectRelationshipMultiResult() throws EntityException, InstantiationException, IllegalAccessException
    {
        QueryCriteria criteria = new QueryCriteria("id", QueryCriteriaOperator.STARTS_WITH, "FIRST ONE").and("child.someOtherField", QueryCriteriaOperator.STARTS_WITH, "HIYA");
        Query query = new Query(AllAttributeForFetch.class, criteria);
        query.setFirstRow(0);
        query.setMaxResults(2);
        query.setQueryOrders(Arrays.asList(new QueryOrder("child.someOtherField"), new QueryOrder("intPrimitive")));
        query.setSelections(Arrays.asList("id", "longValue", "intPrimitive", "child.someOtherField"));

        List results = manager.executeQuery(query);

        Assert.assertNotNull(results);
        Assert.assertTrue(results.size() == 2);
        Assert.assertTrue(results.get(0) instanceof Map);
        Assert.assertTrue(((Map)results.get(0)).get("id").equals("FIRST ONE"));
        Assert.assertTrue(((Map)results.get(1)).get("id").equals("FIRST ONE4"));
    }
}

File: onyx-database-tests/test/memory/queries/SelectQueryTest.kt
package memory.queries

import category.InMemoryDatabaseTests
import com.onyx.exception.OnyxException
import com.onyx.persistence.query.Query
import com.onyx.persistence.query.QueryCriteria
import com.onyx.persistence.query.QueryCriteriaOperator
import com.onyx.persistence.query.QueryOrder
import entities.AllAttributeForFetch
import org.junit.Assert
import org.junit.Test
import org.junit.experimental.categories.Category

import java.util.Arrays

/**
 * Created by timothy.osborn on 1/10/15.
 */
@Category(InMemoryDatabaseTests::class)
class SelectQueryTest : memory.base.PrePopulatedDatabaseTest() {

    @Test
    @Throws(OnyxException::class, InstantiationException::class, IllegalAccessException::class)
    fun testExecuteSelectFields() {
        val criteria = QueryCriteria("stringValue", QueryCriteriaOperator.STARTS_WITH, "Some")
        val query = Query(AllAttributeForFetch::class.java, Arrays.asList("longValue", "intPrimitive"), criteria)
        val results = manager.executeQuery<Any>(query)
        Assert.assertNotNull(results)
        Assert.assertTrue(results.size == 4)
        Assert.assertTrue(results[0] is Map<*, *>)
        Assert.assertTrue((results[0] as Map<*, *>)["longValue"] as Long > 0)
    }

    @Test
    @Throws(OnyxException::class, InstantiationException::class, IllegalAccessException::class)
    fun testSelectOnlyOne() {
        val criteria = QueryCriteria("stringValue", QueryCriteriaOperator.STARTS_WITH, "Some")
        val query = Query(AllAttributeForFetch::class.java, Arrays.asList("longValue", "intPrimitive"), criteria)
        query.firstRow = 0
        query.maxResults = 1
        val results = manager.executeQuery<Any>(query)

        Assert.assertNotNull(results)
        Assert.assertTrue(results.size == 1)
        Assert.assertTrue(results[0] is Map<*, *>)

        if (results[0] is Long) {
            Assert.assertTrue((results[0] as Map<*, *>)["longValue"] as Long > 0)
        } else if (results[0] is Int) {
            Assert.assertTrue((results[0] as Map<*, *>)["longValue"] as Int > 0)
        }
    }

    @Test
    @Throws(OnyxException::class, InstantiationException::class, IllegalAccessException::class)
    fun testSelectTwoOrderBy() {
        val criteria = QueryCriteria("stringValue", QueryCriteriaOperator.STARTS_WITH, "Some")
        val query = Query(AllAttributeForFetch::class.java, Arrays.asList("id", "stringValue", "longValue", "intPrimitive"), criteria)
        query.firstRow = 2
        query.maxResults = 2
        query.queryOrders = Arrays.asList(QueryOrder("stringValue"), QueryOrder("intPrimitive"))

        val results = manager.executeQuery<Any>(query)

        Assert.assertNotNull(results)
        //        Assert.assertTrue(results.size() == 2);
        Assert.assertTrue(results[0] is Map<*, *>)
        Assert.assertTrue((results[0] as Map<*, *>)["intPrimitive"] as Int == 4)
        Assert.assertTrue((results[1] as Map<*, *>)["intPrimitive"] as Int == 3)
    }

    @Test
    @Throws(OnyxException::class, InstantiationException::class, IllegalAccessException::class)
    fun testNoSelect() {
        val criteria = QueryCriteria("stringValue", QueryCriteriaOperator.STARTS_WITH, "Some")
        val query = Query(AllAttributeForFetch::class.java, criteria)
        query.firstRow = 0
        query.maxResults = 2
        query.queryOrders = Arrays.asList(QueryOrder("stringValue"), QueryOrder("intPrimitive"))

        val results = manager.executeQuery<Any>(query)

        Assert.assertNotNull(results)
        Assert.assertTrue(results.size == 2)
        Assert.assertTrue(results[0] is AllAttributeForFetch)
        Assert.assertTrue((results[0] as AllAttributeForFetch).id == "FIRST ONE")
        Assert.assertTrue((results[1] as AllAttributeForFetch).id == "FIRST ONE2" || (results[1] as AllAttributeForFetch).id == "FIRST ONE1")
    }

    @Test
    @Throws(OnyxException::class, InstantiationException::class, IllegalAccessException::class)
    fun testSelectRelationship() {
        val criteria = QueryCriteria("stringValue", QueryCriteriaOperator.STARTS_WITH, "Some")
                .and("child.someOtherField", QueryCriteriaOperator.STARTS_WITH, "HIYA")
        val query = Query(AllAttributeForFetch::class.java, criteria)
        query.firstRow = 0
        query.maxResults = 2
        query.queryOrders = Arrays.asList(QueryOrder("child.someOtherField"), QueryOrder("intPrimitive"))
        query.selections = Arrays.asList("longValue", "intPrimitive", "child.someOtherField")

        val results = manager.executeQuery<Any>(query)

        Assert.assertNotNull(results)
        Assert.assertTrue(results.size == 1)
        Assert.assertTrue(results[0] is Map<*, *>)
    }

    @Test
    @Throws(OnyxException::class, InstantiationException::class, IllegalAccessException::class)
    fun testSelectRelationshipMultiResult() {
        val criteria = QueryCriteria("id", QueryCriteriaOperator.STARTS_WITH, "FIRST ONE").and("child.someOtherField", QueryCriteriaOperator.STARTS_WITH, "HIYA")
        val query = Query(AllAttributeForFetch::class.java, criteria)
        query.firstRow = 0
        query.maxResults = 2
        query.queryOrders = Arrays.asList(QueryOrder("child.someOtherField"), QueryOrder("intPrimitive"))
        query.selections = Arrays.asList("id", "longValue", "intPrimitive", "child.someOtherField")

        val results = manager.executeQuery<Any>(query)

        Assert.assertNotNull(results)
        Assert.assertTrue(results.size == 2)
        Assert.assertTrue(results[0] is Map<*, *>)
        Assert.assertTrue((results[0] as Map<*, *>)["id"] == "FIRST ONE")
        Assert.assertTrue((results[1] as Map<*, *>)["id"] == "FIRST ONE4")
    }
}
